var/spack/repos/builtin/packages/adios2/package.py

```diff
@@ -72,33 +76,24 @@ class Adios2(CMakePackage):
     # DataMan needs dlopen
     conflicts('+dataman', when='~shared')
 
-    # BPFile engine was emulated via a ADIOS1 engine prior to v2.3.0
-    conflicts('+adios1', when='@2.3.0:')
-
     depends_on('cmake@3.6.0:', type='build')
-    depends_on('pkgconfig', type='build', when='@2.2.0:')
-    # The included ffs requires bison and flex but using them makes
-    # the build fail due to an undefined reference.
-    # depends_on('bison', type='build', when='@2.2.0:')
-    # depends_on('flex', when='@2.2.0:')
+    depends_on('pkgconfig', type='build')
 
     depends_on('mpi', when='+mpi')
     depends_on('zeromq', when='+dataman')
+    depends_on('zeromq', when='+ssc')
 
     depends_on('hdf5', when='+hdf5')
     depends_on('hdf5+mpi', when='+hdf5+mpi')
-    depends_on('adios', when='@:2.2.99 +adios1')
-    depends_on('adios+mpi', when='@:2.2.99 +adios1+mpi')
 
     depends_on('c-blosc', when='@2.4.0: +blosc')
-    depends_on('bzip2', when='+bzip2')
+    depends_on('bzip2', when='@2.4.0: +bzip2')
```

Comment:
Rather than use version restrictions in the variant dependency specification, better would be to  express it as a conflict.  Keeping it in `depends_on` lets the variant be enabled for older versions when it shouldn't be allowed.
```python
# support for LibFoo came in 2.4
variant('foo', default=True, description='Support LibFoo')
conflicts('+foo', when='@:2.3')
depends_on('libfoo', when='+foo')
```

Reply:
@chuckatkins will do, thanks!

Reply:
@chuckatkins on a second look, expressing it as a conflict prevents keeping the default as `True` for a variant that was introduced after a certain version. I think this is the best compromise between keeping defaults as True for simple, well-established packaged, e.g.` bzip2` and a seamless installation for any version, e.g. the +bzip2 default won't have any effect on `spack install adios2@2.3.1`. Adding a conflict will force to keep the default as False.